syntax = "proto3";
import "google/protobuf/empty.proto";

message UserLoginRequest {
    string user_name = 1;
}

message OperationResponse {
    bool is_success = 1;
    string msg = 2;
    string ip_address = 3;
    int32 port_number = 4;
}

message UserConnectionRequest {
    string user_name = 1;
    string chat_id = 2;
}

message ChatMessageRequest {
    string from_user = 1;
    string chat_message = 2;
}

message GroupSubscription {
    string chat_id = 1;
}

service ChatOperations {
    rpc UserLogin (UserLoginRequest) returns (OperationResponse) {}
    rpc ConnectUser (UserConnectionRequest) returns (OperationResponse) {}
    rpc DisconnectUser (UserConnectionRequest) returns (OperationResponse) {}
    rpc Disconnect (UserConnectionRequest) returns (google.protobuf.Empty) {}
    rpc SubscribeToGroup (GroupSubscription) returns (OperationResponse) {}
    rpc SendGroupMessage (ChatMessageRequest) returns (OperationResponse) {}
    rpc DiscoverUsers (UserLoginRequest) returns (OperationResponse) {}
}

service ClientOperations {
    rpc ReceiveChatMessage (ChatMessageRequest) returns (OperationResponse) {}
}